### PR TITLE
Fix announcements of our own gossip

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This fixes three separate issues that, together, make announcements of our own `channel_update` messages incredibly unreliable.

### CODE REVIEW

File: lightning-background-processor/src/lib.rs [hunk 1]
```diff
@@ -491,9 +491,10 @@ mod tests {
...
-				match $node.write(&mut $expected_bytes) {
-					Ok(()) => {
-						loop {
+				loop {
+					$expected_bytes.clear();
+					match $node.write(&mut $expected_bytes) {
+						Ok(()) => {
 							match std::fs::read($filepath) {
 								Ok(bytes) => {
 									if bytes == $expected_bytes {
```
Comment (valentinewallace) [inline lightning-background-processor/src/lib.rs:500]: When I was debugging the CI hang, I got as far as figuring out that this is the line that it loops on repeatedly (unless it's repeatedly erroring on `fs::read`). So, I think this patch would fix CI (though IDK what problem that might indicate in the `FilesystemPersister`)

Comment (TheBlueMatt) [reply lightning-background-processor/src/lib.rs:500]: At least for the purposes of this PR, the issue is that the `ChannelManager` can now change out from under us - the `ChannelManager` changes by a byte or two after one or two timer ticks, and thus the original version written at the top of the loop isn't necessarily the version that gets written to disk. It's not actually an issue, though.

File: lightning/src/ln/msgs.rs [hunk 2]
```diff
@@ -711,6 +711,10 @@ pub enum ErrorAction {
 	/// The peer did something harmless that we weren't able to meaningfully process.
 	/// If the error is logged, log it at the given level.
 	IgnoreAndLog(logger::Level),
+	/// The peer provided us with a gossip message which we'd already seen. In most cases this
+	/// should be ignored, but it may result in the message being forwarded if it is a duplicate of
+	/// our own channel announcements.
+	IgnoreDuplicateGossip,
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs:717]: Hmm, its used in a field titled `action` in the error itself. Feels a bit say to say the `action` to be taken on this error is `DuplicateGossip`.

**Review by valentinewallace — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
Rebased.

File: lightning/src/routing/network_graph.rs [hunk 3]
```diff
@@ -1082,8 +1084,10 @@ impl NetworkGraph {
 				macro_rules! maybe_update_channel_info {
 					( $target: expr, $src_node: expr) => {
 						if let Some(existing_chan_info) = $target.as_ref() {
-							if existing_chan_info.last_update >= msg.timestamp {
+							if existing_chan_info.last_update > msg.timestamp {
 								return Err(LightningError{err: "Update older than last processed update".to_owned(), action: ErrorAction::IgnoreAndLog(Level::Gossip)});
+							} else if existing_chan_info.last_update == msg.timestamp {
+								return Err(LightningError{err: "Update had same timestamp as last processed update".to_owned(), action: ErrorAction::IgnoreDuplicateGossip});
```
Comment (arik-so) [inline lightning/src/routing/network_graph.rs:1090]: I don't know if such a comment would be helpful to anybody but me, but I had to look up to verify that the timestamp is strictly monotonic.

**Review by arik-so — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
@valentinewallace could you re-review the rebased version here?

**Review by valentinewallace — APPROVED** [review]

**Pull request merged**